### DISCUSSION TITLE
PyBricks USB program download

### DISCUSSION THREAD

**Discussion by mihaitutu** [pybricks/support#709] [General]
Before everything I want to say that I have studied a lot about PyBricks and I understand why Bluetooth is used to communicate with hubs. Yet there is a certain need for downloading programs via USB cable connection.

I am a robotics teacher with more the 10 teams that participate yearly in robotics competitions such as World Robot Olympiad and First Lego League. These kids win a lot because they are very passionate about Lego and very good programmers. Now, seeing the Spike and Robot Inventor kits, they want to move on from EV3. Partially because it was retired but also because they want a new challenge to learn and program in python. **The main problem is that in these kind of competitions, Bluetooth connections are not allowed during participation**. They are obliged to use the USB cable for downloading the programs. 

In this regard, is there any way to make the download of the programs possible through the USB cable? **Is there something we can do so that we can use PyBricks and compile the code via USB?** 

To prevent some questions, I will give some information upfront:
- Lego has python support but their firmware is lazy. Starting a program means it will run it around 7 seconds after the start button was pressed. As I said, we want to use it in competitions, so time is of an essence...
- Lego's helpbook inside their app is not complete, there are lots of functions which are hidden. We found out about them studying their firmware, but teaching new stuff to kids should not mean we have to dig so deep for main functionalities...

Thank you very much for your time and support!
Mihai

**Comment by dlech** [comment]
I don't know about WRO, but I have participated in FLL for many years. Bluetooth is only prohibited during the robot matches, but not at other times, so I don't think there is a problem with downloading the program with Bluetooth in the pit area as long as during the match you don't use Bluetooth and instead start the program with the button on the hub. Maybe your local regulations are different?

**Reply by mihaitutu** [reply]
You are right, this is happening in our regional / national competition. I can even explain why. Let's talk about WRO in this case. For each category, there are general rules. For RoboMissions (in the past Regular) there is article 5.11 : 

> Bluetooth, Wi-Fi or any remote connection must be switched off during check time and robot runs. Only teams can use remote connections if there is no other way to transfer the code from a device (e.g. a tablet) to the controller. However, it is strongly recommended to transfer code via cable to avoid problems (e.g. multiple devices with the same name) at the competition day. Of course, it is not allowed to interfere or obstruct any other team or robot with the remote connections a team uses. 

So, in this case you are right, we can use Bluetooth in all other moments of the competitions, except check time and robot runs. But, because regionals or nationals have other formats than those stated in the rules, it is not certain when is the check time. This is accepted, as stated in article 10.3:

> The Host Country of the WRO International Final can decided together with WRO on a slightly different format (e.g. different time/number of practice time/rounds), but need to inform all teams no later than 10 weeks before the event on the competition schedule. 

So, the WRO association give a green light to some different format but does not offer clear guidance on how 5.11 applies. Hence the main problem for national organisers.  So, the national organiser hides under another article, found in the specific rules of each age group mission, under heading 6:

> This is why WRO Association recommends our National Organizers decide if they want to adapt the rules for events in their country. They can make the challenges easier for local, regional, and national events, so that all participants have a positive experience. Our National Organizers can make their own choices, so each competition fits their specific situation and ideas. Here we provide some ideas to make the challenges easier.

Using the above, they state no Bluetooth is allowed. And even if I tried to explain this to them, they said that I should use another firmware which accepts USB connection. Something similar happened in FLL contest. 

As a conclusion, in order for my teams to qualify for the International stages, they need to pass the regional / national. And that's how it goes. So, this is why I am asking about this feature...

_PS: I have talked to other colleagues competing in these robotics competitions in other countries and they have encountered such problems as well. Some convinced the national organizer to allow Bluetooth connection in some moments of the stage, others have accepted the ruling._

**Reply by laurensvalk** [reply]
Thanks for the additional details. If I understand correctly, Bluetooth is not allowed at the competition table and during checks. But does that always mean that USB is required? In other words, would it be possible to just run the previously loaded program, or is re-programming required during the competition?

When you run a (previously downloaded) Pybricks program without being connected to a computer, Bluetooth is not used. I’m wondering whether this is satisfactory?

It does scan for connections before you start the program, but you could work around that by starting your program before hand, and make it wait for you to press a button.

Curious to hear your take on this approach, both in terms of the rules and how practical this would be.

**Reply by mihaitutu** [reply]
During such competitions, teams have practice time in which they can adjust the code. After this period they have the official round. So, because of that, they can adjust the code inside the competition and need a USB connection.

I agree with you this is not really in the spirit of evolution, but I am here with this request after more then 6 months and an official competition in which the organizers will not change their mind. So, if I cannot make them change this rule, which is indeed kinda stupid, I need to go around them and achieve my goal: use PyBricks but with an USB connection.

**Reply by laurensvalk** [reply]
Using USB might be possible one day, but I’m afraid it isn’t a top priority for us at the moment. We’re a small team, so we try to focus on features that the majority would like to see. Even if we did work on it, it probably wouldn’t be released during this competition season.

It could change if more users request USB support or if an organization like WRO wants to [sponsor development of such functionality](https://github.com/sponsors/pybricks), but I don’t know if this is likely.

**Reply by mihaitutu** [reply]
That is really sad. Seems like we are stuck in not using PyBricks or not using the new Spike hub for competitions. We will look for other alternatives...

**Comment by laurensvalk** [comment]
Separately from my other response about usb/Bluetooth, I was curious about a few more things.

First of all, thanks for trying Pybricks and sharing your findings. Feedback like this really helps!

> Lego has python support but their firmware is lazy. Starting a program means it will run it around 7 seconds after the start button was pressed.

Does this happen with every program? Or perhaps mainly with bigger ones? I’m curious to see an example.

>  they want to move on from EV3. Partially because it was retired but also because they want a new challenge to learn and program in python.

For what it’s worth, Pybricks is also available for EV3 and it supports USB. On SPIKE Pybricks will run faster though.

**Reply by mihaitutu** [reply]
> Does this happen with every program? Or perhaps mainly with bigger ones? I’m curious to see an example.

Using Lego app and Lego firmware will just not do. I can write a simple program or a very complex one, it does not matter. There is that delay from the button press to program initialization. When I say a simple program, I mean run a loop with motor start power. When I mean a complex program, I mean a basic PD controller for line follower. 

> For what it’s worth, Pybricks is also available for EV3 and it supports USB. On SPIKE Pybricks will run faster though.

Before PyBricks was an option, we found that the only viable solution for competition was EV3Basic (https://sites.google.com/site/ev3basic/). We are happy with what we developed so far and maybe, if we get PyBricks working on all hubs, we will move full on it.

**Comment by butch01** [comment]
Hi there, are there any news / progress about the possibility to upload code via usb to a pybricks device (in my case robot inventor hub). Thanks.

**Reply by laurensvalk** [reply]
We're very close to getting this ready, but we have to figure out some connection issues for Windows before we can release it.

**Comment by paravmellanox** [comment]
@laurensvalk ,

3 of my teams need to use pybricks in the WRO 2026. Regional and nationals team do not allow bluetooth connections whats so ever during the competitions. We must need USB connectivity.

When can we expect this support and in which version of the pybricks? Can you please update the plan, so we can decide to use/not-use pybricks.

Ability to download or not via USB is single decision factor here.

**Reply by laurensvalk** [reply]
We should be able to release a preview version with USB support this month!

**Reply by paravmellanox** [reply]
@laurensvalk , great. Looking forward to it. Would like to try as soon as available.

**Reply by paravmellanox** [reply]
@laurensvalk ,
Hi,

Its been few months. Do we have USB download working with Pybricks for Lego spike prime?

**Reply by laurensvalk** [reply]
Yes! You can use this version which includes USB: https://beta.pybricks.com/
